File: 0291-word-pattern-ii/0291-word-pattern-ii.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

class Solution {
    public boolean wordPatternMatch(String pattern, String s) {
        Map<Character, String> mapping = new HashMap<>();
        Set<String> mapped = new HashSet<>();

        boolean result = helper(0, 0, pattern, s, mapping, mapped);
        
        return result;
    }

    private boolean helper(int i, int j, String pattern, String s, Map<Character, String> mapping, Set<String> mapped) {
        if (i == pattern.length() && j == s.length()) {
            return true;
        }

        if (i >= pattern.length() || j >= s.length()) {
            return false;
        }

        if (!mapping.containsKey(pattern.charAt(i))) {
            for (int length = 1; length <= s.length() - j; length++) {
                String substring = s.substring(j, j + length);
                if (!mapped.contains(substring)) {
                    mapped.add(substring);
                    mapping.put(pattern.charAt(i), substring);
                    if (helper(i + 1, j + length, pattern, s, mapping, mapped)) {
                        return true;
                    }
                    mapping.remove(pattern.charAt(i));
                    mapped.remove(substring);
                }
            }
        } else {
            String mappedString = mapping.get(pattern.charAt(i));
            if (s.startsWith(mappedString, j) && helper(i + 1, j + mappedString.length(), pattern, s, mapping, mapped)) {
                return true;
            }
        }

        return false;
    }
}

File: 0291-word-pattern-ii/0291-word-pattern-ii.py
class Solution(object):
    def wordPatternMatch(self, pattern, s):
        """
        :type pattern: str
        :type s: str
        :rtype: bool
        """
        mapping = {} # keep track of pattern to s mapping 
		# keep track of what s is already being mapped 
		# so you can avoid mapping two different pattern to same thing 
        mapped = set() 
        
        def helper(i, j): 
            if i == len(pattern) and j == len(s): 
                return True 
            if i >= len(pattern) or j >= len(s): # did not work, backtrack
                return 
            if pattern[i] not in mapping: 
                for length in range(1, len(s)-j+1): 
                    if s[j:j+length] not in mapped: # can't map to a mapped string
                        mapped.add(s[j:j+length])
                        mapping[pattern[i]] = s[j:j+length]
                        if helper(i+1, j+length): 
                            return True 
                        mapping.pop(pattern[i])
                        mapped.remove(s[j:j+length])
			# current pattern[i] exists, see if it s[j:] starts with it
            elif s[j:j+len(mapping[pattern[i]])] == mapping[pattern[i]]:  
                if helper(i+1, j+len(mapping[pattern[i]])):
                    #print(mapping)
                    return True
            return False 
        
        x = helper(0, 0)
        print(mapping)
        return x